### PR TITLE
feat(dropbox): add tpa sync

### PR DESCRIPTION
**Pull request by athielba**
## Description

In the PR, the following feature have been included

1. fetch all team linked apps and send to elba
2. third party apps
    1. refresh-object webhook & refresh-object inngest handler
    2. delete-object webhook & delete-object inngest handler
    3. trigger-scan webhook  & trigger-scan inngest handler
## Checklist:

Before you create this PR, confirm all the requirements listed below by checking the checkboxes like this (`[x]`).

- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests to cover the new feature or fixes.

### CODE REVIEW

File: apps/dropbox/src/app/api/webhooks/elba/third-party-apps/refresh-object/service.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      organisationId,
+      userId: userId,
+      appId,
+      isFirstSync: true,
+    },
+  });
+
+  return {
+    success: true,
+  };
```
Comment (elbalexandre) [inline apps/dropbox/src/app/api/webhooks/elba/third-party-apps/refresh-object/service.ts:33]:
```suggestion
```

File: apps/dropbox/src/app/api/webhooks/elba/third-party-apps/start-sync/route.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import { parseWebhookEventData } from '@elba-security/sdk';
+
+export async function POST(request: Request) {
+  const data: unknown = await request.json();
+
+  const { organisationId } = parseWebhookEventData('third_party_apps.scan_triggered', data);
+
+  const response = await startThirdPartySync(organisationId);
+
+  return NextResponse.json(response);
```
Comment (elbalexandre) [inline apps/dropbox/src/app/api/webhooks/elba/third-party-apps/start-sync/route.ts:12]:
```suggestion
  await startThirdPartySync(organisationId);

  return new NextResponse();
```

File: apps/dropbox/src/inngest/functions/third-party-apps/delete-object.test.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+});
+
+describe('third-party-apps-delete-objects', () => {
+  beforeEach(async () => {
+    await db.delete(organisations);
+    await insertOrganisations({});
+    vi.spyOn(crypto, 'decrypt').mockResolvedValue('token');
+    vi.clearAllMocks();
+  });
+
```
Comment (elbalexandre) [inline apps/dropbox/src/inngest/functions/third-party-apps/delete-object.test.ts:40]: We could test that we got a NonRetriableError when the organisation does not exists

File: apps/dropbox/src/inngest/functions/third-party-apps/delete-object.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  await step.run('delete-third-party-apps-objects', async () => {
+    return await dbx.deleteTeamMemberThirdPartyApp({
+      teamMemberId: userId,
+      appId,
+    });
+  });
+
+  return {
+    success: true,
+  };
```
Comment (elbalexandre) [inline apps/dropbox/src/inngest/functions/third-party-apps/delete-object.ts:33]:
```suggestion
```

File: apps/dropbox/src/inngest/functions/third-party-apps/refresh-objects.test.ts [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+);
+
+describe('third-party-apps-refresh-objects', () => {
+  beforeEach(async () => {
+    await db.delete(organisations);
+    await insertOrganisations({});
+    vi.spyOn(crypto, 'decrypt').mockResolvedValue('token');
+    vi.clearAllMocks();
+  });
+
```
Comment (elbalexandre) [inline apps/dropbox/src/inngest/functions/third-party-apps/refresh-objects.test.ts:42]: There is missing test when the org is not registered in the db.

File: apps/dropbox/src/app/api/webhooks/elba/third-party-apps/delete-object/service.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    data: {
+      organisationId,
+      userId,
+      appId,
+    },
+  });
+
+  return {
+    success: true,
+  };
```
[the remark this answers is not in the document]
Comment (athielba) [reply apps/dropbox/src/app/api/webhooks/elba/third-party-apps/delete-object/service.ts:32]: I will use same as other integrations like,
```
 return {
    status: 'completed',
  };

File: .gitignore [hunk 7]
```diff
@@ -33,4 +33,4 @@ yarn-error.log*
 .turbo
 
 # vercel
-.vercel
+.vercel
```
Comment (elbalexandre) [inline .gitignore:36]:
```suggestion
.vercel

```

File: apps/dropbox/src/env.ts [hunk 8]
```diff
@@ -44,5 +44,11 @@ export const env = z
     TOKEN_REFRESH_MAX_RETRY: zEnvRetry(),
     VERCEL_ENV: z.string().optional(),
     ELBA_REGION: z.string(),
+    DROPBOX_TPA_DELETE_OBJECT_RETRIES: zEnvRetry(),
+    DROPBOX_TPA_DELETE_OBJECT_CONCURRENCY: zEnvInt(),
+    DROPBOX_TPA_REFRESH_OBJECT_RETRIES: zEnvRetry(),
+    DROPBOX_TPA_REFRESH_OBJECT_CONCURRENCY: zEnvInt(),
+    DROPBOX_TPA_SYNC_RETRIES: zEnvRetry(),
+    DROPBOX_TPA_SYNC_CONCURRENCY: zEnvInt(),
```
Comment (elbalexandre) [inline apps/dropbox/src/env.ts:52]: I suggest we add a default value to each of theses env variable so we don't have to set them in Vercel.
@raulelba have a nice version of the `zEnvRetry` that could be added in `@elba-security/inngest` (zod could be added as a peerDeps: `pnpm add --save-peer zod`).

File: apps/dropbox/src/inngest/functions/third-party-apps/refresh-objects.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    await elba.thirdPartyApps.updateObjects({
+      apps,
+    });
+  });
+
+  return {
+    status: 'completed',
+  };
```
Comment (elbalexandre) [inline apps/dropbox/src/inngest/functions/third-party-apps/refresh-objects.ts:60]: The status will be always `completed`. This might be misleading.

Comment (elbalexandre) [reply apps/dropbox/src/inngest/functions/third-party-apps/refresh-objects.ts:60]: Having one step in a function is useless, it's just add noise to the code.
Two solutions:
- remove the step
- add one step to retrieve tpa + one to update/delete the object

File: .gitignore (hunk 7 above)
Comment (athielba) [reply .gitignore:36]: it is annoying , I don't know why it keeps appearing

**Review by elbalexandre — APPROVED** [review]

**Pull request merged**